compiler/luci/pass/src/QuantizeOnnxQDQPass.cpp

```diff
@@ -0,0 +1,237 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "QuantizeOnnxQDQPass.h"
+#include "QuantizationUtils.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+namespace
+{
+
+using namespace luci;
+
+class OnnxQDQPattern final
+{
+public:
+  OnnxQDQPattern(luci::CircleCustomOut *candidate) { dq_out = candidate; }
+
+public:
+  bool matched()
+  {
+    if (not dq_out)
+      return false;
+
+    dq = loco::must_cast<luci::CircleCustom *>(dq_out->input());
+    if (not is_onnx_dequantize_linear(dq))
+      return false;
+
+    q_out = dynamic_cast<luci::CircleCustomOut *>(dq->inputs(0));
+    if (not q_out)
+      return false;
+
+    dq_scale = dynamic_cast<luci::CircleConst *>(dq->inputs(1));
+    if (not dq_scale)
+      return false;
+
+    dq_zerop = dynamic_cast<luci::CircleConst *>(dq->inputs(2));
+    if (not dq_zerop)
+      return false;
+
+    q = loco::must_cast<luci::CircleCustom *>(q_out->input());
+    if (not is_onnx_quantize_linear(q))
+      return false;
+
+    input = loco::must_cast<luci::CircleNode *>(q->inputs(0));
+    if (input->dtype() != loco::DataType::FLOAT32)
+      return false;
+
+    q_scale = dynamic_cast<luci::CircleConst *>(q->inputs(1));
+    if (not q_scale)
+      return false;
+
+    q_zerop = dynamic_cast<luci::CircleConst *>(q->inputs(2));
+    if (not q_zerop)
+      return false;
+
+    const auto q_dtype = q->dtype();
+    const auto q_scale_dtype = q_scale->dtype();
+    const auto q_zerop_dtype = q_zerop->dtype();
+    const auto dq_scale_dtype = dq_scale->dtype();
+    const auto dq_zerop_dtype = dq_zerop->dtype();
+
+    if (q_scale_dtype != loco::DataType::FLOAT32)
+      return false;
+
+    if (dq_scale_dtype != loco::DataType::FLOAT32)
+      return false;
+
+    // Invariant from onnx Quantize operator
+    if (q_dtype != q_zerop_dtype)
+      return false;
+
+    // Invariant from onnx Dequantize operator
+    if (q_dtype != dq_zerop_dtype)
+      return false;
+
+    // Check length of scale, zp = 1
+    if (q_scale->size<loco::DataType::FLOAT32>() != 1)
+      return false;
+
+    if (dq_scale->size<loco::DataType::FLOAT32>() != 1)
+      return false;
+
+    auto q_zerop_size = 0;
+    auto dq_zerop_size = 0;
+    switch (q_zerop_dtype)
+    {
+      case loco::DataType::S16:
+        q_zerop_size = q_zerop->size<loco::DataType::S16>();
+        dq_zerop_size = dq_zerop->size<loco::DataType::S16>();
+        break;
+      default:
+        throw std::runtime_error("Unsupported zerop dtype in " + q_zerop->name());
+    }
+
+    if (q_zerop_size != 1)
+      return false;
+
+    if (dq_zerop_size != 1)
+      return false;
+
+    return true;
+  }
+
+public:
+  luci::CircleCustomOut *dq_out = nullptr;
```

Comment:
(optional) for readability
1/ use struct as `struct OnnxQDQPattern final` or
2/ change member with `_` previx like `_dq_out`

Reply:
I've followed the first suggestion. PTAL